### PR TITLE
wip: Discussions

### PR DESCRIPTION
**Pull request by nicksnyder**
This is not mergeable for a few reasons:
- [ ] uses proposed api that is not finalized (we will need to publish our extension as preview or create a separate preview extension for dogfooding, https://sourcegraph.slack.com/archives/C0LJP8WTB/p1537195769000100)
- [x] There is some hard-coded access tokens for local development
- [ ] I am not sure of all the semantics of the revisions we pass to Sourcegraph. Needs vetting by @slimsag 
- [ ] Does not handle local modifications to files (should limit commenting range to non-modified ranges?)
- [ ] Add logic to refresh known threads periodically

### CODE REVIEW

**Comment by emidoots** [comment]
> uses proposed api that is not finalized (I think we can't publish with this, unless maybe we mark this extension as preview? Unclear.)

My hope is that we can get this working behind e.g. a feature flag and at least ship to everyone on our team. I'll try to figure out if this is possible or not

**Comment by emidoots** [comment]
> Does not handle local modifications to files (should limit commenting range to non-modified ranges?)

I'll think about this some more.

For reading existing comments, it should be easy for us to support placing the comments in the right locations on modified files (for now, we can do this by sending the code in your editor to the sourcegraph server. In the future, before we ship to any real users, this would be done client side using the same line resolution logic).

For creating new comments, the biggest risk is probably just making sure that we clearly denote "I am commenting on some modified code" in the DB so we don't end up with a lot of garbage that we later need to filter out somehow on web. I'll think of how to do this.

**Comment by nicksnyder** [comment]
The [GitHub Pull Requests](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github) extension got published with proposed api, so unsure if they got a special exception or if we just need to mark our extension as [preview](https://github.com/Microsoft/vscode-pull-request-github/blob/master/package.json#L14)

**Comment by emidoots** [comment]
I will push a commit to unify some of the GraphQL logic here to share it a bit more with https://github.com/sourcegraph/sourcegraph-extension-discussions

**Comment by emidoots** [comment]
https://github.com/sourcegraph/sourcegraph/pull/13253 will address the whole repo resolution issue, will push a change here soon to implement it on this end

File: src/graphql.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+'use strict'
```
[the remark this answers is not in the document]
Comment (felixfbecker) [reply src/graphql.ts]: You dont have to copy paste this code. We could share it. Some things in here are also unneeded, like the symbol-backed gql tag.
We could generate functions for queries and mutations, but if you need to specify what fields you want

**Comment by felixfbecker** [comment]
I agree, the reason GraphQL clients are usually _not_ generated is that the GraphQL query language itself is seen as the "interface" or "SDK" for APIs. A folder that must be kept in sync sounds like it causes more trouble than it offers benefits.

**Comment by felixfbecker** [comment]
cc @dadlerj who worked on GraphQL error handling

File: src/shared/api.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            }
+          }
+        }
+        ${discussionThreadFieldsFragment(relativeRev)}
+        ${discussionCommentFieldsFragment}
+      `,
+        { input, relativeRev }
+    )
+    if (!data || !data.discussions || !data.discussions.createThread) {
+        throw createAggregateError(errors)
```
Comment (dadlerj) [inline src/shared/api.ts:107]: We have nicer error handling now using `dataOrThrowErrors`! See https://sourcegraph.sgdev.org/github.com/sourcegraph/sourcegraph/-/blob/src/backend/graphql.tsx#L21:17 and examples in https://sourcegraph.sgdev.org/github.com/sourcegraph/sourcegraph/-/commit/78e0791ca3e3d4ac37d13b38c6da84473cb6c4e5

File: src/shared/api.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+              }
+            }
+          }
+        }
+        ${discussionThreadFieldsFragment(opts.relativeRev)}
+      `,
+        opts
+    )
+    if (!data || !data.discussionThreads) {
+        throw createAggregateError(errors)
```
Comment (dadlerj) [inline src/shared/api.ts:167]: Same here

File: src/shared/api.ts [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+      `,
+        { threadID, relativeRev }
+    )
+    if (
+        !data ||
+        !data.discussionThreads ||
+        !data.discussionThreads.nodes ||
+        data.discussionThreads.nodes.length !== 1
+    ) {
+        throw createAggregateError(errors)
```
Comment (dadlerj) [inline src/shared/api.ts:206]: and here

File: src/shared/api.ts [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            }
+          }
+        }
+        ${discussionThreadFieldsFragment(relativeRev)}
+        ${discussionCommentFieldsFragment}
+      `,
+        { threadID, contents, relativeRev }
+    )
+    if (!data || !data.discussions || !data.discussions.addCommentToThread) {
+        throw createAggregateError(errors)
```
Comment (dadlerj) [inline src/shared/api.ts:242]: and here

File: src/shared/api.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    const { data, errors } = await queryGraphQL(
+        `
+        query RenderMarkdown($markdown: String!, $options: MarkdownOptions) {
+          renderMarkdown(markdown: $markdown, options: $options)
+        }
+      `,
+        { markdown }
+    )
+    if (!data || !data.renderMarkdown) {
+        throw createAggregateError(errors)
```
Comment (dadlerj) [inline src/shared/api.ts:262]: and here

**Comment by emidoots** [comment]
I've thought about this a lot more over the past few hours. Here is my final conclusion / stance on the topic:

- I am actually OK with duplicating the GraphQL queries / fragments / function definitions and having three separate copies. I'm not happy about it, and I think it is ugly / dumb that GraphQL inherently forces us to do this, but I am OK with this because there isn't a better _generalizable approach_.

But, I strongly dislike the fact that we have separate error handling and GraphQL request code that is copied, sprinkled, and modified throughout several of our codebases, and Dan's feedback above just re-enforces my stance here that it is impossible to know which file to copy.

To be specific, why does the browser extension still use `createAggregateError`? How do we know the version of `graphql.ts` copied into this repo by Nick is the correct version? Why must all of these implementations be different and each contain their own bugs or imperfections in their own various ways?:

1. `sourcegraph-vscode`: Nick's [`graphql.ts`](https://github.com/sourcegraph/sourcegraph-vscode/blob/22dd51e4276baa4ca47a8e4647fb336efe058478/src/graphql.ts) file here that was obviously copied from _some other implementation_ (who knows where?).
2. `sourcegraph-extension-discussions`: The [`graphql.ts`](https://github.com/sourcegraph/sourcegraph-extension-discussions/blob/c5634c71355f28c0d2d519d641db8df7e6a04727/src/graphql.ts) file I wrote for the Sourcegraph Extension, which sucks / doesn't support the `gql` marker syntax because implementing these files is hard.
3. `sourcegraph/sourcegraph`: The [`graphql.tsx`](https://sourcegraph.sgdev.org/github.com/sourcegraph/sourcegraph/-/blob/src/backend/graphql.tsx#L21:17) file we have in the webapp repo which apparently has better error handling today than `createAggregateError` according to Dan above.
4. `extensions-client-common`: The [`graphql.ts`](https://github.com/sourcegraph/extensions-client-common/blob/83765793abe93dd0378f04f3c2b29255ce577abd/src/graphql.ts) file we have in the extensions-client-common repo, which is apparently using its own beautiful fork of [`createAggregateError`](https://github.com/sourcegraph/extensions-client-common/blob/8376579/src/errors.ts#L35).
5. `browser-extensions`: Yet another [`graphql.tsx`](https://github.com/sourcegraph/browser-extensions/blob/master/src/shared/backend/graphql.tsx) and [`errors.tsx`](https://github.com/sourcegraph/browser-extensions/blob/master/src/shared/backend/errors.tsx) fork. Is it better than the webapp's implementation? Why doesn't it use `extensions-client-common`'s version?

I think this is a problem worth solving before we merge this PR for the following reasons:

1. I encountered issues with the error handling here as I stated earlier.
2. With this PR, we introduce yet another copy, and with `sourcegraph-extension-discussions` we introduce yet another copy. The problem is rapidly growing and getting out of hand.
3. Even if we don't update all of the codebases, it seems like this problem would be easy to solve by publishing just one implementation here as an npm package (_*if* we knew which one was the best one!_).

**Comment by felixfbecker** [comment]
There are also open source solutions out there that we could and probably should use. I will just need some time to evaluate them.

**Comment by emidoots** [comment]
The short answer is: I don't know. I don't have the answers here. I am just calling out what is from my perspective a major pain point in using GraphQL in our codebase, that I feel should be easily addressable somehow.

- I tried to address this by sharing the code between this extension and `sourcegraph-extension-discussions` but it is hard because in the case of Sourcegraph Extensions the GraphQL requests have to be done through a VS Code-like command API not through fetch, so I didn't know how to publish this as an npm package easily. And my approach obviously wasn't generalizable to the other codebases.
- Writing `sourcegraph-extension-discussions`, I didn't know how to write these `queryGraphQL`, `mutateGraphQL` functions and I couldn't simply copy the others. Figuring out how to write these properly would've taken me several hours (and writing the entire extension itself only took me 5, so that is not acceptable).
- Working on this PR the error handling just didn't work, and the codebase was completely different than other codebases I've worked on with GraphQL error handling.  So I guess I should've debugged what was wrong in this PR instead of trying to share the code(?) but I feel like that would have again taken me a few hours to figure out and that time was better spent elsewhere.

**Comment by nicksnyder** [comment]
I am getting this graphql error: "invalid graphql.ID" when creating a new thread.

```
mutation CreateThread($input: DiscussionThreadCreateInput!, $relativeRev: String!) {
          discussions {
            createThread(input: $input) {
              ...DiscussionThreadFields
              comments {
                totalCount
                nodes {
                  ...DiscussionCommentFields
                }
              }
            }
          }
        }
        
    fragment DiscussionThreadFields on DiscussionThread {
      id
      author {
        ...UserFields
      }
      title
      target {
        __typename
        ... on DiscussionThreadTargetRepo {
          repository {
            name
          }
          path
          branch {
            displayName
          }
          revision {
            displayName
          }
          selection {
            startLine
            startCharacter
            endLine
            endCharacter
            linesBefore
            lines
            linesAfter
          }
          
        relativePath(rev: $relativeRev)
        relativeSelection(rev: $relativeRev) {
          startLine
          startCharacter
          endLine
          endCharacter
        }
    
        }
      }
      inlineURL
      createdAt
      updatedAt
      archivedAt
    }

    fragment UserFields on User {
      displayName
      username
      avatarURL
    }
  
        
    fragment DiscussionCommentFields on DiscussionComment {
      id
      author {
        ...UserFields
      }
      contents
      html
      inlineURL
      createdAt
      updatedAt
    }
```

```
{input: {title: "asd", contents: "asd", targetRepo: {repositoryGitCloneURL: "git@github.com:nicksnyder/go-i18n", path: "v2/i18n/bundle.go", branch: "master", …}}, relativeRev: "master"}
```

**Comment by nicksnyder** [comment]
> sourcegraph-vscode: Nick's graphql.ts file here that was obviously copied from some other implementation (who knows where?).

The bulk of this file is requestGraphQL, which is not copied because I had to make it work in a node environment. The convenience functions and gql tag were copied, but there is not much there.

> And debugging this wasn't trivial for me because the error handling implementation was 100% different than the other two implementations.

I spent no time thinking about or dealing with edge/error cases (this was hack quality). I suspect it will be hard to share error code because the right way to treat errors in a VS Code extension may be different than web.

**Comment by emidoots** [comment]
@nicksnyder Odd, that must mean that the `repositoryGitCloneURL` bit is failing for some reason but I have no clue why that would be or why it would be producing such a lame error. Are you running the latest `sourcegraph/sourcegraph` master branch with my change for `repositoryGitCloneURL` merged in?

**Comment by emidoots** [comment]
@nicksnyder that's totally fair :) I just feel we gotta get error handling working properly before merging this

**Comment by nicksnyder** [comment]
I am running sourcegraph/enterprise, do I need to run sourcegraph/sourcegraph?

**Comment by emidoots** [comment]
no, you should run `$ENTERPRISE/dev/start.sh` but make sure you have an up to date copy of sourcegraph/sourcegraph cloned in your gopath

**Comment by nicksnyder** [comment]
>  make sure you have an up to date copy of sourcegraph/sourcegraph cloned in your gopath

Oops, this was my problem 😬

**Comment by emidoots** [comment]
Moving to https://github.com/sourcegraph/sourcegraph-vscode/pull/19

**Pull request closed without merging**